verl/experimental/agent_loop/tool_agent_loop.py

```diff
@@ -60,170 +107,299 @@ def init_class(cls, config, tokenizer, processor, **kwargs):
-                if tool_response.image or tool_response.video:
-                    # Multi-modal content with structured format
-                    content = []
-                    if tool_response.image:
-                        content.append({"type": "image"})
-                    if tool_response.video:
-                        content.append({"type": "video"})
-                    if tool_response.text:
-                        content.append({"type": "text", "text": tool_response.text})
-                    message = {"role": "tool", "content": content}
-                else:
-                    # Text-only content
-                    message = {"role": "tool", "content": tool_response.text or ""}
+    async def _handle_generating_state(self, agent_data: AgentData, sampling_params: dict[str, Any]) -> AgentState:
+        """Handle the generating state: generate model response and check for tool calls."""
+        # Clear add_messages to ensure it only contains messages from this turn
+        agent_data.add_messages.clear()
 
-                tool_messages.append(message)
+        with simple_timer("generate_sequences", agent_data.metrics):
+            output = await self.server_manager.generate(
+                request_id=agent_data.request_id,
+                prompt_ids=agent_data.prompt_ids,
+                sampling_params=sampling_params,
+                image_data=agent_data.image_data,
+            )
+
+        agent_data.response_ids = output.token_ids
+        agent_data.prompt_ids += agent_data.response_ids
+        agent_data.response_mask += [1] * len(agent_data.response_ids)
+        if output.log_probs:
+            agent_data.response_logprobs += output.log_probs
+
+        # Check termination conditions
+        if len(agent_data.response_mask) >= self.response_length:
+            return AgentState.TERMINATED
+
+        # Extract tool calls
+        _, agent_data.tool_calls = await self.tool_parser.extract_tool_calls(agent_data.response_ids)
+
+        # Handle interaction if needed
+        if self.interaction_config_file:
+            assistant_message = await self.loop.run_in_executor(
+                None, lambda: self.tokenizer.decode(agent_data.response_ids)
+            )
+            agent_data.add_messages.append({"role": "assistant", "content": assistant_message})
 
-                # Handle image data
+        # Determine next state
+        if agent_data.tool_calls:
+            return AgentState.PROCESSING_TOOLS
+        elif self.interaction_config_file:
+            return AgentState.INTERACTING
+        else:
+            return AgentState.TERMINATED
+
+    async def _handle_processing_tools_state(self, agent_data: AgentData) -> AgentState:
+        """Handle the processing tools state: execute tool calls and prepare tool responses."""
+        # Clear add_messages to ensure it only contains messages from this turn
+        agent_data.add_messages.clear()
```

Comment:
Please use a local variable to track messages of this turn.

Reply:
I still don't get why we need `add_messages` field in agent_data since we're clear it in each turn.